0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,283 @@
+
+- `swap`
+	- id
+    - alpha_ledger_fk -> protocol_id
+    - beta_ledger_fk -> protocol_id
+    - ...
+
+- `han_ethereum_swap_details`
+	- id
+	- wei_quantity
+	- ...
+
+----
+
+This approach does not give us full type safety as data stored in `han_ethereum_swap_details` is not bound to the swap at all.
+
+4)
+
+[source]
+----
+
+
+- `swap`
+	- id
+    - ...
+
+- `han_ethereum_swap_details`
+	- id
+    - swap_id_fk -> swap_id
+	- wei_quantity
+    - ledger : (alpha | beta)
+	- ...
+
+----
+
+Eventually we decided to go for the last approach as this is the least amount of added tables reduces the join complexity when loading data.
+
+This allows us to :
+
+* add additional protocols easily (one new table per protocol, e.g. `{protocol}_swap_detail`)
+* rename protocol tables as needed :)
+
+The second thing which was a bit challenging was how to model data which is only known at a later point in time and how to model data which not all `Swaps` have in common.
+We decided to go for an _extension table_ approach which allows us to add additional information where needed:
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `local_swap_id` `role` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/86bfc97bcea91602eaf3e40e6da7d71beaff9911/assets/db_schema/han_bitcoin_swap_protocols_without_state.puml&fmt=svg[HanBitcoinProtocols]
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
```

Comment:
I think I commented somewhere that nullable columns for the identities might be a good idea as it gets rid of these extension tables. Did this get lost when you worked in the feedback or did you decide against it?

Reply:
yes, this was feedback was lost 😇

Reply:
updated